File: app/src/main/java/com/rockspin/bargainbits/watch_list/WatchedItem.java
/*
 * Copyright (c) Rockspin 2015.
 *
 * All Rights Reserved
 */

package com.rockspin.bargainbits.watch_list;

import android.os.Parcelable;

import com.rockspin.bargainbits.data.models.cheapshark.CompactDeal;
import com.rockspin.bargainbits.data.models.GameSearchResult;
import com.rockspin.bargainbits.data.models.cheapshark.GameInfo;

import java.io.Serializable;

import auto.parcel.AutoParcel;

@AutoParcel
public abstract class WatchedItem implements Parcelable, Serializable {
    static final long serialVersionUID = 1L;

    public abstract String gameName();
    public abstract String gameId();
    public abstract float watchedPrice();

    public boolean hasCustomWatchPrice() {
        return watchedPrice() >= 0;
    }

    public static WatchedItem create(String gameName, String gameId, float watchedPrice) {
        return new AutoParcel_WatchedItem(gameName, gameId, watchedPrice);
    }

    public static WatchedItem from(GameSearchResult game){
        return  create(game.getName(), game.getGameID(), (float) game.getCheapestPrice());
    }

    public static WatchedItem from(CompactDeal compactDeal){
        return  create(compactDeal.getGameName(),compactDeal.getGameId(), compactDeal.getLowestSalePrice());
    }

    public static WatchedItem from(GameInfo gameInfo) {
        return create(gameInfo.getInfo().getGameName(), gameInfo.getInfo().getGameId(), gameInfo.getLowestSalePrice());
    }
}

File: app/src/main/java/com/rockspin/bargainbits/watch_list/WatchedItem.kt
/*
 * Copyright (c) Rockspin 2015.
 *
 * All Rights Reserved
 */

package com.rockspin.bargainbits.watch_list

import android.os.Parcel
import android.os.Parcelable
import com.rockspin.bargainbits.data.models.GameSearchResult
import com.rockspin.bargainbits.data.models.cheapshark.CompactDeal
import paperparcel.PaperParcel
import java.io.Serializable

@PaperParcel
data class WatchedItem(
    val gameName: String,
    val gameId: String,
    // TODO - change to Double without breaking old watch list entries?
    val watchedPrice: Float
) : Parcelable, Serializable {

    val hasCustomWatchPrice: Boolean
        get() = watchedPrice >= 0

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        PaperParcelWatchedItem.writeToParcel(this, dest, flags)
    }

    companion object {
        @JvmField val CREATOR = PaperParcelWatchedItem.CREATOR

        internal const val serialVersionUID = 1L

        fun from(game: GameSearchResult): WatchedItem {
            return WatchedItem(game.name, game.gameID, game.cheapestPrice.toFloat())
        }

        fun from(compactDeal: CompactDeal): WatchedItem {
            return WatchedItem(compactDeal.gameName, compactDeal.gameId, compactDeal.lowestSalePrice)
        }
    }
}
